src/service/builder.js

```diff
@@ -0,0 +1,236 @@
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {LayoutPriority} from '../layout';
+import {READY_SCAN_SIGNAL} from './resources-interface';
+import {VisibilityState} from '../visibility-state';
+import {getServiceForDoc, registerServiceBuilderForDoc} from '../service';
+import {hasNextNodeInDocumentOrder, isIframed} from '../dom';
+import {removeItem} from '../utils/array';
+
+const ID = 'builder';
+
+/** @implements {../service.Disposable} */
+export class Builder {
+  /** @param {!./ampdoc-impl.AmpDoc} ampdoc  */
+  constructor(ampdoc) {
+    /** @private @const */
+    this.ampdoc_ = ampdoc;
+
+    const {win} = ampdoc;
+
+    /** @private @const {!IntersectionObserver} */
+    this.observer_ = new win.IntersectionObserver((e) => this.observed_(e), {
+      // Root bounds are not important, so we can use the `root:null` for a
+      // top-level window.
+      root: isIframed(win) ? win.document : null,
+      rootMargin: '250% 31.25%',
+      threshold: 0.001,
+    });
+
+    /** @private @const {!Map<!AmpElement, {asap: boolean, isIntersecting: boolean}>} */
+    this.targets_ = new Map();
+
+    /** @private {?Array<!AmpElement>} */
+    this.parsingTargets_ = [];
+
+    /** @private {boolean} */
+    ampdoc.whenReady().then(() => this.checkParsing_());
+
+    /** @private {?UnlistenDef} */
+    this.visibilityUnlisten_ = ampdoc.onVisibilityChanged(() =>
+      this.docVisibilityChanged_()
+    );
+  }
+
+  /** @override */
+  dispose() {
+    this.observer_.disconnect();
+    this.targets_.clear();
+    if (this.visibilityUnlisten_) {
+      this.visibilityUnlisten_();
+      this.visibilityUnlisten_ = null;
+    }
+  }
+
+  /**
+   * @param {!AmpElement} target
+   */
+  scheduleAsap(target) {
+    this.targets_.set(target, {asap: true, isIntersecting: false});
+    this.waitParsing_(target);
+  }
+
+  /**
+   * @param {!AmpElement} target
+   */
+  schedule(target) {
+    if (this.targets_.has(target)) {
+      return;
+    }
+
+    if (target.deferredBuild()) {
+      this.targets_.set(target, {asap: false, isIntersecting: false});
+      this.observer_.observe(target);
+    } else {
+      this.targets_.set(target, {asap: false, isIntersecting: true});
+    }
+
+    this.waitParsing_(target);
+    this.signalScanReady_();
+  }
+
+  /**
+   * @param {!AmpElement} target
+   */
+  unschedule(target) {
+    if (!this.targets_.has(target)) {
+      return;
+    }
+
+    this.targets_.delete(target);
+
+    this.observer_.unobserve(target);
+
+    if (this.parsingTargets_) {
+      removeItem(this.parsingTargets_, target);
+      this.checkParsing_();
```

Comment:
I don't believe this call is necessary here? Nothing about the remaining pending elements should have changed, right?

Reply:
It's not strictly necessary. But generally what it does is trying to check parsing state more often - whenever we know the DOM changed. And unscheduling happens when DOM changes. I can remove, but it likely provides some, however minor, benefit.

Reply:
FMI: what are all of the points at which we check parsing state? When onDocumentReady, when any item is scheduled, and when any item is unscheduled?

It seems reasonable either way to me. 
Why can't CE have a `parsedCallback` 🙄

Reply:
CE don't have `parsedCallback` because elements might be created manually, e.g. using `e = createElement('x'); body.append(e); e.append(someChild)` - the order of such creations are completely undefined. So the spec authors have preferred mutability. Some (for now most) of our elements cannot rebuild themselves when DOM changes, so this is critical for us at this time. However, even for elements that can rebuild themselves - it's still better to wait for performance reasons then keep rebuilding itself while the element is still being parsed.

RE: when we schedule. It's best to think about it this way:
1. We decide that an element might not be yet fully parsed. We can only make that judgment until `DOMReady`. And we don't actually know that - we just think it might be the case. We look if there are any siblings after this element. If there are some - we know for certain that the element has been parsed. If not - maybe/maybe not.
2. Since we're not sure and we have to look for "next sibling" we need to keep checking. We can only check when we know more DOM has been parsed. Or in our case, some other elements have been added or removed.

So any time a new element has been scheduled or unscheduled is a good time for us to check if some of the pending elements have been "completed".